examples/cfq/models.py

```diff
@@ -0,0 +1,219 @@
+    positions are ignored as indicated by the mask. This is also called
+    "Bahdanau attention" and was originally proposed in:
+    ```
+    Bahdanau et al., 2015. Neural Machine Translation by Jointly Learning to
+    Align and Translate. ICLR. https://arxiv.org/abs/1409.0473
+    ```
+
+    Args:
+      query: The query with which to compute attention. Shape:
+        <float32>[batch_size, 1, query_size].
+      projected_keys: The inputs multiplied with a weight matrix. Shape:
+        <float32>[batch_size, seq_length, attention_size].
+      values: The values to be weighted [batch_size, seq_length, values_size]
+      mask: A mask that determinines which values in `keys` are valid. Only
+        values for which the mask is True will get non-zero attention scores.
+        <bool>[batch_size, seq_length].
+      hidden_size: The hidden size of the MLP that computes the attention score.
+
+    Returns:
+      The weighted values (context vector) [batch_size, seq_len]
+    """
+    # The projected_keys could be cached, since typically multiple keys are
+    # compared with the same projected_keys.
+    # projected_keys = nn.Dense(keys, hidden_size, name='keys', bias=False)
+    projected_query = nn.Dense(query, hidden_size, name='query', bias=False)
+    # Query broadcasts in the sum below along the time (seq_length) dimension.
+    energy = nn.tanh(projected_keys + projected_query)
+    scores = nn.Dense(energy, 1, name='energy', bias=False)
+    scores = scores.squeeze(-1)  # New shape: <float32>[batch_size, seq_len].
+    # TODO: see if I can rewrite this to not have the squeezing and unsqueezing
+    scores = jnp.where(mask, scores, -jnp.inf)  # Using exp(-inf) = 0 below.
+    scores = nn.softmax(scores, axis=-1)
+
+    attention = jnp.sum(jnp.expand_dims(scores,2) * values, axis=1)
+    return attention  # Shape: <float32>[batch_size, seq_len]
+
+class Decoder(nn.Module):
+    """LSTM decoder."""
+
+    def apply(self, 
+              init_state,
+              encoder_hidden_states,
+              attention_mask,
+              inputs: jnp.array,
+              shared_embedding: nn.Module,
+              vocab_size: int,
+              teacher_force: bool = False):
+        """Apply decoder model"""
+        lstm_cell = nn.LSTMCell.shared(name='lstm')
+        projection = nn.Dense.shared(features=vocab_size, name='projection')
+        mlp_attention = MlpAttention.partial(
+                                       hidden_size=ATTENTION_SIZE).shared(name = 'attention')
+        projected_keys = nn.Dense(encoder_hidden_states, ATTENTION_SIZE, name='keys', bias=False)
+
+        def decode_step_fn(carry, x):
+            rng, (c,h), last_prediction = carry
+            carry_rng, categorical_rng = jax.random.split(rng, 2)
+            if not teacher_force:
+                x = last_prediction
+            x = shared_embedding(x)
+            dec_prev_state = jnp.expand_dims(h,1)
```

Comment:
Why do we have to add a dimension here?

Reply:
Because the attention module takes the query of shape [batch_size, 1, query_size] and before expand it has shape [batch_size, query_size] (where query_size = lstm hidden_size)